aerospike-core/src/cluster/partition_tokenizer.rs

```diff
@@ -23,65 +22,50 @@ use crate::cluster::Node;
 use crate::commands::Message;
 use crate::errors::{ErrorKind, Result};
 use crate::net::Connection;
-use aerospike_rt::RwLock;
 
 const REPLICAS_NAME: &str = "replicas-master";
 
 // Validates a Database server node
 #[derive(Debug, Clone)]
 pub struct PartitionTokenizer {
     buffer: Vec<u8>,
-    length: usize,
-    offset: usize,
 }
 
 impl PartitionTokenizer {
     pub async fn new(conn: &mut Connection) -> Result<Self> {
         let info_map = Message::info(conn, &[REPLICAS_NAME]).await?;
         if let Some(buf) = info_map.get(REPLICAS_NAME) {
             return Ok(PartitionTokenizer {
-                length: info_map.len(),
                 buffer: buf.as_bytes().to_owned(),
-                offset: 0,
             });
         }
         bail!(ErrorKind::BadResponse("Missing replicas info".to_string()))
     }
 
-    pub async fn update_partition(
+    pub fn update_partition(
         &self,
-        nmap: Arc<RwLock<HashMap<String, Vec<Arc<Node>>>>>,
+        nmap: &mut HashMap<String, [Option<Arc<Node>>; node::PARTITIONS]>,
         node: Arc<Node>,
-    ) -> Result<HashMap<String, Vec<Arc<Node>>>> {
-        let mut amap = nmap.read().await.clone();
-
+    ) -> Result<()> {
         // <ns>:<base64-encoded partition map>;<ns>:<base64-encoded partition map>; ...
         let part_str = str::from_utf8(&self.buffer)?;
-        let mut parts = part_str.trim_end().split(|c| c == ':' || c == ';');
-        loop {
-            match (parts.next(), parts.next()) {
-                (Some(ns), Some(part)) => {
+        for part in part_str.trim_end().split(';') {
+            match part.split_once(':') {
+                Some((ns, part)) => {
                     let restore_buffer = base64::decode(part)?;
-                    match amap.entry(ns.to_string()) {
-                        Vacant(entry) => {
-                            entry.insert(vec![node.clone(); node::PARTITIONS]);
-                        }
-                        Occupied(mut entry) => {
-                            for (idx, item) in entry.get_mut().iter_mut().enumerate() {
-                                if restore_buffer[idx >> 3] & (0x80 >> (idx & 7) as u8) != 0 {
-                                    *item = node.clone();
-                                }
-                            }
+                    let entry = nmap.entry(ns.to_string()).or_insert_with(||[(); node::PARTITIONS].map(|_|None));
```

Comment:
Start with an empty partition table, so we can tell that each partition has been claimed by at least one node or at least notify the caller.